### PR TITLE
Add cmdlet for Resolving Hyak and AutoRest errors

### PR DESCRIPTION
**Pull request by markcowl**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using System.Collections;
+using Microsoft.Azure.Commands.ResourceManager.Common;
+
+namespace Microsoft.Azure.Commands.Profile.Errors
+{
+    [Cmdlet(VerbsDiagnostic.Resolve, "AzureRmError", DefaultParameterSetName = ResolveError.AnyErrorParameterSet)]
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs]: @markcowl does it make sense to have an alias to `Resolve-Error`?

File: src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    [OutputType(typeof(AzureErrorRecord))]
+    [OutputType(typeof(AzureExceptionRecord))]
+    [OutputType(typeof(AzureRestExceptionRecord))]
+    public class ResolveError : AzureRMCmdlet
+    {
+        public const string AnyErrorParameterSet = "AnyErrorParameterSet";
+        public const string LastErrorParameterSet = "LastErrorParameterSet";
+        [Parameter(Mandatory =false, Position =0, HelpMessage ="The error records to resolve", ValueFromPipeline =true, ParameterSetName = AnyErrorParameterSet)]
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs]: @markcowl nit: can we put an empty line between the last parameter set and the start of this parameter?

File: src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        public ErrorRecord[] Error { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "The last error", ParameterSetName = LastErrorParameterSet)]
+        public SwitchParameter Last { get; set; }
+        protected override IAzureContext DefaultContext
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs]: @markcowl nit: can we put an empty line between the last parameter and this property?

File: src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            {
+                foreach (var error in records)
+                {
+                    ErrorRecord record = error as ErrorRecord;
+                    HandleError(record);
+                }
+            }
+        }
+
+        IEnumerable<ErrorRecord> GetErrorVariable()
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs]: @markcowl nit: should the below methods be private if they're only being used for this cmdlet?

File: src/ResourceManager/Profile/Commands.Profile/help/Resolve-AzureRmError.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+external help file: Microsoft.Azure.Commands.Profile.dll-Help.xml
+online version: 
+schema: 2.0.0
+---
+
+# Resolve-AzureRmError
+
+## SYNOPSIS
+
+Display detailed information about Azure PowerShell errors.
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/help/Resolve-AzureRmError.md]: @markcowl are the errors specific to Azure PowerShell?

File: src/ResourceManager/Profile/Commands.Profile/Errors/AzureErrorRecord.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            InvocationInfo = record.InvocationInfo;
+            ScriptStackTrace = record.ScriptStackTrace;
+            ErrorCategory = record.CategoryInfo;
+            ErrorDetails = record.ErrorDetails;
+        }
+
+        public ErrorDetails ErrorDetails { get; set; }
+
+        public ErrorCategoryInfo ErrorCategory { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/AzureErrorRecord.cs]: @markcowl nit: can we put an empty line between `ErrorCategory` and `InvocationInfo`?

File: src/ResourceManager/Profile/Commands.Profile/Errors/AzureExceptionRecord.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        public AzureExceptionRecord(Exception exception, ErrorRecord record, bool inner = false) : base(record)
+        {
+            Message = exception.Message;
+            HelpLink = exception.HelpLink;
+            StackTrace = exception.StackTrace;
+            Exception = exception;
+        }
+
+        public bool InnerException { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/AzureExceptionRecord.cs]: @markcowl nit: can we put an empty line between `InnerException` and `Exception`?

File: src/ResourceManager/Profile/Commands.Profile/Errors/HttpRequestInfo.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+        public HttpRequestInfo(HttpRequestMessageWrapper request) : base(request)
+        {
+            Uri = request.RequestUri;
+            Verb = request.Method.ToString();
+        }
+
+        public Uri Uri { get; set; }
+        public string Verb { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Profile/Commands.Profile/Errors/HttpRequestInfo.cs]: @markcowl nit: can we put an empty line between `Uri` and `Verb`?

File: src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs (hunk 4 above)
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile/Errors/ResolveError.cs]: methods with no visibility specified are automatically private

File: src/ResourceManager/Profile/Commands.Profile/help/Resolve-AzureRmError.md (hunk 5 above)
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile/help/Resolve-AzureRmError.md]: No, we do special processing for azure powershell errors, but it is explained in the description that it applies to all errors as well

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/job/powershell-demand/1593/

Want to make sure there are no problems when generating help. LGTM once this succeeds.

**Comment by markcowl** [comment]
@cormacpayne Is this the whitespace issue again?

**Comment by markcowl** [comment]
new on-deemand run: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1599/

**Pull request merged**